### PR TITLE
Adds subDir checkout for git components

### LINKED ISSUE #3933: Re-enable support for the sparseCheckoutDir field in devfiles
/kind bug
/area devfile
/area regression

## What versions of software are you using?

**Operating System:**
macOS

**Output of `odo version`:**
master

## How did you run odo exactly?
`odo create java-openliberty --starter`

## Actual behavior
When cloning a starter project with the `sparseCheckoutDir` field set, `odo create --starter` now clones the entire source code repo specified in the starter project, rather than checking out the directory specified in `sparseCheckoutDir`.

As the java-openliberty devfile currently uses `sparseCheckoutDir` for its starterProject, we're currently unable to use `odo create java-openliberty --starter` and successfully deploy a component from it.

## Expected behavior
`odo create --starter` respects the `sparseCheckoutDir` field in devfiles

## Any logs, error output, etc?
Sample devfile:
```
schemaVersion: 2.0.0
metadata:
  name: nodejs
  version: 1.0.0
starterProjects:
  - name: nodejs-starter
    git:
      remotes:
        origin: "https://github.com/odo-devfiles/nodejs-ex.git"
      sparseCheckoutDir: /test
components:
  - name: runtime
    container:
      image: registry.access.redhat.com/ubi8/nodejs-12:1-45
      memoryLimit: 1024Mi
      mountSources: true
      sourceMapping: /project
      endpoints:
        - name: http-3000
          targetPort: 3000
commands:
  - id: install
    exec:
      component: runtime
      commandLine: npm install
      workingDir: /project
      group:
        kind: build
        isDefault: true
  - id: run
    exec:
      component: runtime
      commandLine: npm start
      workingDir: /project
      group:
        kind: run
        isDefault: true
```

### PR DESCRIPTION
**Pull request by mik-dass**
Signed-off-by: mik-dass <mrinald7@gmail.com>

**What type of PR is this?**

/kind feature

**What does does this PR do / why we need it**:

Updates schema to change sparseCheckoutDir to subDir and adds support of it for git starter projects.

**Which issue(s) this PR fixes**:

Fixes #3933 

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

- Create a devfile with subDir set and place it in a empty folder.
- Run `odo create --starter`
- The starter project's `subDir` content should be downloaded only and the temp folder should be removed.

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -1248,3 +1248,166 @@ func DisplayLog(followLog bool, rd io.ReadCloser, compName string) (err error) {
...
+		err = fs.RemoveAll(filepath.Join(originalPath, file.Name()))
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+// GitSubDir handles subDir for git components
+func GitSubDir(srcPath, destinationPath, subDir string, fs filesystem.Filesystem) error {
```
Comment (girishramnani) [inline pkg/util/util.go:1367]: well you shouldn't be exposing functions that have `fs` in them as the filesystem argument only changes when testing. 
So you should have a `gitSubDir` that has the `fs` and then the `GitSubDir` should call `gitSubDir(...., filesystem.DefaultFs{})`

Comment (girishramnani) [reply pkg/util/util.go:1367]: this should happen at other places too

**Review by girishramnani — CHANGES_REQUESTED** [review]
^

**Comment by dharmit** [comment]
> **How to test changes / Special notes to the reviewer**:
> 
>   * Create a devfile with subDir set and place it in a empty folder.
> 
>   * Run `odo create --starter`
> 
>   * The starter project's `subDir` content should be downloaded only and the temp folder should be removed.

I'm a bit lost as to what I should be expecting when testing the functionality. So what I did was:
* Use the devfile mentioned in the https://github.com/openshift/odo/issues/3933#issue-698700370
* Run `odo create --starter`

What I observed is:
* Here's what is cloned:
  ```sh
  $ ls -al
  total 24
  drwxr-xr-x.  3 dshah dshah 4096 Nov  3 13:12 .
  drwxr-xr-x. 33 dshah dshah 4096 Nov  3 13:10 ..
  drwxr-x---.  3 dshah dshah 4096 Nov  3 13:15 .odo
  -rw-r--r--.  1 dshah dshah  801 Nov  3 13:12 devfile.yaml
  -rw-------.  1 dshah dshah   34 Nov  3 13:14 .gitignore
  -rw-r--r--.  1 dshah dshah  492 Nov  3 13:12 test.js
  ```

  It's missing other files in the https://github.com/odo-devfiles/nodejs-ex repo
* Doing `odo push` works without any error but the URL doesn't load. I haven't looked into the underlying services or pods yet.

File: pkg/util/util.go (hunk 1 above)
Comment (mik-dass) [reply pkg/util/util.go]: Fixed

**Comment by mik-dass** [comment]
> It's missing other files in the https://github.com/odo-devfiles/nodejs-ex repo

Since the `subDir` is defined as `/test`, the only file copied was the `test.js`. Rest of the files are ignored as they are outside the `test` directory.

> Doing odo push works without any error but the URL doesn't load. I haven't looked into the underlying services or pods yet.

The url doesn't work since you defined the subDir as `test` and thus only the test related file was copied. `odo push` passes as nodejs doesn't complain during the build.

**Comment by kadel** [comment]
> I'm a bit lost as to what I should be expecting when testing the functionality. So what I did was:
> 
> * Use the devfile mentioned in the [#3933 (comment)](https://github.com/openshift/odo/issues/3933#issue-698700370) and changed `sparseCheckoutDir` to `subDir`:
>   ```
>   $ grep -A5 -B5 subDir devfile.yaml
>   metadata:
>     name: nodejs
>     version: 1.0.0
>   starterProjects:
>     - name: nodejs-starter
>       subDir: /test
>       git:
>         remotes:
>           origin: "https://github.com/odo-devfiles/nodejs-ex.git"
>   components:
>     - name: runtime
>   ```
> * Run `odo create --starter`
> 
> What I observed is:
> 
> * Here's what is cloned:
>   ```shell
>   $ ls -al
>   total 24
>   drwxr-xr-x.  3 dshah dshah 4096 Nov  3 13:12 .
>   drwxr-xr-x. 33 dshah dshah 4096 Nov  3 13:10 ..
>   drwxr-x---.  3 dshah dshah 4096 Nov  3 13:15 .odo
>   -rw-r--r--.  1 dshah dshah  801 Nov  3 13:12 devfile.yaml
>   -rw-------.  1 dshah dshah   34 Nov  3 13:14 .gitignore
>   -rw-r--r--.  1 dshah dshah  492 Nov  3 13:12 test.js
>   ```
>   
>   
>   It's missing other files in the https://github.com/odo-devfiles/nodejs-ex repo
> * Doing `odo push` works without any error but the URL doesn't load. I haven't looked into the underlying services or pods yet.

@dharmit This is correct and expected behavior, This is how it is supposed to work.

**Comment by girishramnani** [comment]
we need to get this PR merged asap so I am lgtming it. Tests are good and the code quality is great so we are good to go.
/lgtm

**Pull request merged**